crypto/mlkem/build.info

```diff
@@ -1,3 +1,3 @@
 LIBS = ../../libcrypto
 
-SOURCE[../../libcrypto] = mlkem768.c
+SOURCE[../../libcrypto] = mlkem.c mlkem512.c mlkem768.c mlkem1024.c
```

Comment:
BTW @slontis in his ML-DSA and SLH-DSA PRs consistently uses `ml_dsa` and `slh_dsa` for file names. Should we rename the files here?

Reply:
fwiw, I'm not a big fan of "underscore" names as they're a nuisance to type. But then again, who types any more these days? AI does it all for us...

Reply:
I don't much care about the file names, but indeed requiring the use of the shift-key is not developer friendly, so if there's no compelling reason to use `_` I tend to not go there.

Reply:
The APIs and names **should** have the underscore in them as a separator and we should be **consistent**.

NIST explicitly uses ML-KEM not MLKEM (or mlkem) so we should not have dropped the logical separator between the two parts of the name in my view. Mapping of hyphen to underscore is a common approach.

I would land that change in now rather than delaying it as frankly people need to get used to the "right" naming.

Having the file names consistent is a good idea too - and using underscore in our file names is very much our standard approach (1596 files with underscores, 646 without) so we do use underscore as a logical separator in file names.

Reply:
@t-j-h So, for these functions (purely internal to our default, and some day FIPS providers) non-public interfaces that get little-o "ossl_" prefixes, to keep the consumers of the static ".a" library out of namespace issues, but otherwise their names nobody's business, I take it the suggestion is:

1. Separate the `ml` and the `kem` with an `_` as in `ml_kem_512_encode_public_key`, et. al.
2. Do the same with the file names, `ml_kem.c`, `ml_kem_512.c`, ...

Now I don't have a strong view on "2.", except that it is a bit of a nuisance.  But "1" sucks, because you probably also want to see associated variables renamed, error messages updated, ... 

The TLS group names [don't have separators](https://datatracker.ietf.org/doc/html/draft-connolly-tls-mlkem-key-agreement-05#section-4.1), just `mlkem512`, ... Correspondingly the [IANA registry](https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-8) also lists flattened names.

I'm increasingly inclined to author a header file with abbreviated names for all these, and use short aliases in the code, with only the compiled code having whatever the latest fashion dictates. :-)

Reply:
fwiw, I'm also all for consistency. Does it exist in OpenSSL right now? Do users create "rsa-2048", "rsa_2048" or "rsa2048" keys? "ripemd160WithRSA" is another interesting example. I also fail to find a single instance of "ed_25519" in the code, but lots of "ed25519" instances in APIs and names.

Then there is the above-mentioned difference between the IETF TLS group name (MLKEM768) and the NIST KEM algorithm name (ML-KEM-768). Maintaining that difference _will_ cause a big FAQ entry at the very least ("For some algorithms you cannot use the name returned by `openssl list -kem-algorithms` for specifying the algorithm as a TLS group").

Final argument: What's easier to type? mlk768, ml_kem_768, ml-kem-768? It will (have to) be done gazillions of time when this lands... I'd argue that the answer to this question should drive the decision (for consistency) which name to use in the source code.

Reply:
The people who name the standards and publish the standards get to pick the name. That various standards are not consistent isn't our issue. We shouldn't be adding to the confusion generating different names. You are looking for the one-true-consistent-naming-scheme which unfortunately does not exist. 

What I'm pointing out is we should generally be using the names directly from the standards in the manner in which the standards define them - consistent or not with other standards.

Which is why the IETF specified hybrid has a different approach to the NIST specified underlying algorithm standard.

PKCS#11 and KMIP both consistently use the NIST defined terms for the NIST standards (following the same principle).
For SLH-DSA and ML-DSA there are also the HASH variants to be handled and that is a HASH- prefix to the naming if we following the NIST way of naming things (as the algorithms get different algorithm identifiers for the HASH variants). 

For Ed25519 that is what the author of the specification uses as the terminology - no underscore, no space, in camel case form (just to add to the context).

Reply:
I think that provides the consensus answer - the other NIST algorithms being implemented all follow the principles outlined - and they do make sense - so the principle of use what is in the specification drives things along with the added consistency to make this the same as the other two. I would make the file names consistent as well as it makes sense to me to have them matching when there is no strong reasons to not do so. It is something we really should do.

I also do think we should have things like this written down so time isn't wasted - but it also makes sense to get this in place now so we don't waste time as users start to look at and use the code (and that is the point of a feature branch - to get more feedback and testing in advance of it hitting the main repo). 

It is good to see this code rapidly coming into usable shape!

Reply:
> I would make the file names consistent as well as it makes sense to me to have them matching when there is no strong reasons to not do so. It is something we really should do.

I am not yet seeing the blatant specificity I was asking for. What **exactly** do you think must or as appropriate should be changed.

For example, changing the function names is a lot of work, with little payoff, because they're not seen by users... Must they change?

The file names are likewise an internal artefact. What matters a lot more is what we call the algorithms in documentation, in the various "fetch" APIs, and in the TLS "group" registry, ... those are all user-visible.

I agree it is best (i.e. required) to get the docs and public APIs right. I am much less enthused about having to be absolutely consistent w.r.t. use or non-use of "_" to mirror use or non-use of "-" in the associated external algorithm names when it comes to internal APIs and file names. So long as there's consistency **within** each algorithm implementation.  Nobody will be confused IMHO by whether the file names or symbols are "mlkem" vs. on the other hand "slh_dsa". You quickly figure it out and move on...

Reply:
You are seeing it - and it is clearly stated. But for additional clarity - make them **all** consistent. 

This is an open source project. A portion of our community members spend time in the source. Consistency in the source is important. Taking a position that this is "internal" and no one looks at it and we can just do what we like doesn't match the context we are in. 

There is zero benefit in having a different approach and inconsistencies other than avoiding a **small** one-time cost to change to be consistent. And then everyone looking at the code will see a nicer, cleaner, consistent usage. Maintainability is improved by consistency.

We are trying to raise the bar and improve the state of the library including the state of the source - and unnecessary  inconsistencies do not help with that. 

We are actually looking to improve consistent across algorithms - as already noted - ML-KEM, ML-DSA and SLH-DSA have no reason to be inconsistent.

Reply:
I take it your position is that not only the files, but all he symbol names have to also change.

Reply:
In particular:

- Code: `ossl_mlkem_`*some_method* → `ossl_ml_kem_`*same_method*.
- Code: `ossl_mlkem768_`*some_method* → `ossl_ml_kem_768_`*same_method* (ditto for 512/1024).
- Files: `some/path/mlkem`*whatever* → `some/path/ossl_ml_kem_`*whatever*.

Reply:
+1 from me to being consistent in naming files and symbols where possible.

Reply:
> +1 from me to being consistent in naming files and symbols where possible.

Can you specific about what *consistent* means here?

- There's consistency *within* an algorithm's implementation
- There's consistency across algorithm implementations (though it is not clear to me that whether hyphens are dropped from symbol names and files or replaced with underscores is really a consistency concern when looking at separate algorithms)
- There's consistency with related implementations of the API making the names more predictable when looking across implementations of the same algorithm

Moving the files around is quite cheap, but without also renaming the symbols slightly degrades internal consistency.
Renaming all the symbols is a pain, and feels like unnecessary make-work.

Reply:
Rename everything to be consistent with the names we are using for the algorithm.
External APIs. Internal APIs (where the internal APIs have some form of the algorithm name within them).
File names.
Hyphens map to underscores. That is consistent. The "logical" separator is there in the name and you don't drop that.
The C language (outside of VMS and a few other "interesting" environments) doesn't allow hyphens in symbol names or we would most likely use them rather than mapping hyphen to underscore. 

Do this for every new algorithm that we implement and then we will be consistent with the standards that specify the algorithms (and we didn't have any disagreement that it makes sense to match the standards in the external APIs and anything the user might reference or see).

Then we do this for all new stuff - and it will make ML-KEM, ML-DSA and SLH-DSA consistent. If we have non-consistent stuff in ML-DSA and/or SLH-DSA we will apply the same principles there.

Reply:
Among reasons for the `ossl_` prefix on exposed "internal" symbols, there's wanting to avoid symbol clashes with other libraries.  (I actually recall that as a driving factor that got us started, but maybe I'm biased)

For source file names, I don't see the same technical need...  so I take that it's more about cosmetics/esthetics?  Either way, that one was new to me, I can't recall having heard that "need" before now.

Reply:
> The C language (outside of VMS and a few other "interesting" environments) doesn't allow hyphens in symbol names or we would most likely use them rather than mapping hyphen to underscore.

Note: VMS is no exception in this regard.  The extra character that VMS C allows is the `$` ('cause VMS uses that as a prefix separator for its stuff).

Reply:
> Do users create "rsa-2048", "rsa_2048" or "rsa2048" keys?

I assume the question is rhetorical...

So ML-KEM veers off from previous key type names by including the size (where for others, you get to specify the size, or group, or whatever, as parameters).  I'm not sure what to think of that, but don't feel qualified to argue about it either, but am noting it as a curiosity.

Reply:
They aren't "sizes" as such in the same sense at all - and in the other PQC algorithms they are different things.
Think of them as pointless parameter set variations - and not something the user is permitted to vary at all in a conforming implementation - i.e. fixed NIST choices.

Reply:
The main issue is that one can't change just the "size".  There are three parameter **sets**, these are more analogous to ECDSA **named curves** than to RSA user-configurable key sizes.  The only question is whether there should be a single backend provider that takes the variant parameter set name as a *PARAM*, or three separate providers (method array handles) each specific to one of the variants.

Now it turns out in that fact all but initialisation methods of the backend methods are actually common, so the code reuse is there already. And the difference is basically whether the user fetches "ML-KEM" and then has to specify a variant as a param during key creation, or whether they fetch "ML-KEM-<size>" and don't need to specify anything further (unless they want the derandomised interface).

With those as the options, it is pretty clear that being more explicit during fetch is by far the simpler UI. So the choice is clear.

Reply:
Speaking of *derandomisation*, I am reminded that in this PR presently the functions that can be derandomised have names like:  `ossl_mlkem512_encap_rand`, `ossl_mlkem512_encap_seed`. The second is derandomised. But in mlkem-native, they chose `..._enc` vs. `..._enc_derand`.

Perhaps that's a better choice of terminology, because there a small chance of confusion with the key generation seed, vs. the nonce (a.k.a. *entropy*) for the encapsulation operation? I don't see this as a significant issue, but it is not difficult to change the (non-public, internal to the provider) API to be:

```
..._genkey()
..._genkey_derand()
..._encap()
..._encap_derand()
..._decap()
```
rather than
```
..._genkey_seed()
..._genkey_rand()
..._encap_seed()
..._encap_rand()
..._decap()
```
**if** the alternative is significantly more clear.